base_name_search_multi_lang/models/ir_model.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2020 Ecosoft Co., Ltd (http://ecosoft.co.th/)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+from odoo import api, fields, models, tools
+
+
+@tools.ormcache("model")
+def _get_model_name_search_multi_lang(self, model):
+    multi_lang = self.env["ir.model"].search_read(
+        [("model", "=", self._name)], ["name_search_multi_lang"]
+    )[0]["name_search_multi_lang"]
+    return multi_lang
+
+
+def _extend_name_search_lang(self, lang, name, args, operator, result, limit):
+    res_lang = self.with_context(lang=lang)._name_search(
+        name, args, operator, limit=limit
+    )
+    res_ids = [x[0] for x in result]
+    new_res = filter(lambda x: x[0] not in res_ids, res_lang)
+    result.extend(new_res)
+    return result
+
+
+class IrModel(models.Model):
+    _inherit = "ir.model"
+
+    name_search_multi_lang = fields.Boolean(
+        string="Search Translated Name",
+        help="Name search this model from all translated languages",
+    )
+
+    @api.constrains("name_search_multi_lang")
+    def update_name_search_multi_lang(self):
+        self.clear_caches()
+
+    def _register_hook(self):
+        def make_name_search():
+            @api.model
+            def name_search(self, name="", args=None, operator="ilike", limit=100):
+                res = name_search.origin(
+                    self, name=name, args=args, operator=operator, limit=limit
+                )
+                # For model with name_search_multi_lang, extend result
+                multi_lang = _get_model_name_search_multi_lang(self, self._name)
+                if multi_lang:
+                    context_lang = self._context.get("lang")
+                    installed_langs = self.env["res.lang"].get_installed()
+                    langs = [x[0] for x in installed_langs if x[0] != context_lang]
+                    for lang in langs:
+                        res = _extend_name_search_lang(
```

Comment:
Looks like this will exceed the limit easily, as well as the default model search order. Would it help to finish with a  regular id search on collected ids with the given limit, and then return name_get on the result of that?

Reply:
You mean limit of the results, i.e., 8 records?

Reply:
Yes, from the limit argument of name_search. You apply the limit in every search, but then you keep extending the result set. And extending the result set that way does not keep the ordering as defined on the model.

Reply:
I have made a bunch of sample data. It surprised me that, even with .extend, it does show more than 7 rows. None the less, I re-browse again this should ensure the order.